os/drivers/pwm.c

```diff
@@ -445,27 +445,40 @@ static int pwm_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
 
 	case PWMIOC_SETCHARACTERISTICS: {
 		FAR const struct pwm_info_s *info = (FAR const struct pwm_info_s *)((uintptr_t)arg);
-		DEBUGASSERT(info != NULL && lower->ops->start != NULL);
+		DEBUGASSERT(info != NULL);
 
 #ifdef CONFIG_PWM_PULSECOUNT
 		pwmvdbg("PWMIOC_SETCHARACTERISTICS frequency: %d duty: %08x count: %d started: %d\n", info->frequency, info->duty, info->count, upper->started);
 #else
 		pwmvdbg("PWMIOC_SETCHARACTERISTICS frequency: %d duty: %08x started: %d\n", info->frequency, info->duty, upper->started);
 #endif
-
-		/* Save the pulse train characteristics */
-
-		memcpy(&upper->info, info, sizeof(struct pwm_info_s));
-
 		/* If PWM is already running, then re-start it with the new characteristics */
-
 		if (upper->started) {
+			DEBUGASSERT(lower->ops->stop != NULL);
+			ret = lower->ops->stop(lower);
+			if (ret < 0) {
+				return ret;
+			}
+			upper->started = false;
 #ifdef CONFIG_PWM_PULSECOUNT
-			ret = lower->ops->start(lower, &upper->info, upper);
-#else
-			ret = lower->ops->start(lower, &upper->info);
+			if (upper->waiting) {
```

Comment:
As I know basic flow is below,
open -> ioctl(PWMIOC_SETCHARACTERISTICS) -> ioctl(PWMIOC_START)..
pwm_start is being called by ioctl(PWMIOC_START), and please notice that this flag is only for upper's state..

Reply:
The upper->waiting is used for CONFIG_PWM_PULSECOUNT mode. It means that if the number of pulse does not reach to the specified value (that is, the upper still waits the pulse hit the specified counter value,) it can set the characteristics by stopping the original, but restart the next pulse with new parameters. Thanks for the comment.

Reply:
One thing to note is that
PWM characteristics can be changed during the start phase, that is;
open -> set char -> start -> set char (stop->set char->restart) ->...

Reply:
I mean it is only for upper's state, and it will clear when it expired or stopped.
So I think lower should handle callback mechanism to clear this.

Reply:
Got your point. As you said, when CONFIG_PWM_PULSECOUNT is set, pwm_start() will waits until pwm_expired() callback is called. That is the right process. However, If other app or thread wants to make PWM stop, it can call ioctl to stop it, therefore upper can directly control here.